### PR TITLE
Adding a dependency check in mocha

### PR DESCRIPTION
**Pull request by gauntface**
@surma @addyosmani 

I have some concerns that this will be a bit harsh on PR's etc, but at the same time I want something that forces dependency updates. Thoughts?

### CODE REVIEW

**Comment by robwierzbowski** [comment]
IMO a test error is one step too opinionated. There are valid reasons that
an org may want to use an earlier version of a dependency (especially if
it's a front end bundled dependency).

On Friday, February 5, 2016, Matthew Gaunt <notifications@github.com
<javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:

> @surma <https://github.com/surma> @addyosmani
> <https://github.com/addyosmani>
>
> I have some concerns that this will be a bit harsh on PR's etc, but at the
> same time I want something that forces dependency updates. Thoughts?
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/google/web-starter-kit/pull/827
> Commit Summary
>
>    - Adding a dependency check in mocha tests
>
> File Changes
>
>    - *M* package.json
>    <https://github.com/google/web-starter-kit/pull/827/files#diff-0> (1)
>    - *A* test/.eslintrc
>    <https://github.com/google/web-starter-kit/pull/827/files#diff-1> (5)
>    - *A* test/dependency-checker.js
>    <https://github.com/google/web-starter-kit/pull/827/files#diff-2> (74)
>
> Patch Links:
>
>    - https://github.com/google/web-starter-kit/pull/827.patch
>    - https://github.com/google/web-starter-kit/pull/827.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/web-starter-kit/pull/827>.
>


-- 
Rob Wierzbowski
@robwierzbowski <http://twitter.com/#!/robwierzbowski>
http://github.com/robwierzbowski
http://robwierzbowski.com

**Comment by gauntface** [comment]
I agree it feels a bit aggressive but I'm also happy to skip versions and a comment - same as I'm doing with gulp 4.0. 

Feel like its something worth trying and remove if it becomes a pain.

**Comment by surma** [comment]
To be honest, I *love* it. If somebody wants an older version of a package, that package should be *explicitly* whitelisted so people know that the author knows about the out-of-date package. 

That’s what I would like to see added to this PR: A file with a list of package names that are excluded from this check (which could also include `gulpjs/gulp#4.0`). WDYT?

File: test/dependency-checker.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  it('should have up to date npm dev-dependencies', done => {
+    const packageManifest = require(path.join(__dirname, '..', 'package.json'));
+    david.getUpdatedDependencies(packageManifest, {dev: true, stable: true}, function(er, deps) {
+      if (er) {
+        return done(er);
+      }
+
+      const outdatedDependencies = Object.keys(deps);
+      outdatedDependencies.forEach((dependencyName, index) => {
+        if (!dependencyCaveat[dependencyName]) {
```
Comment (surma) [inline test/dependency-checker.js]: Why are we only doing this for `devDependencies`? This is just as legit for actual dependencies, isn’t it?

From a coding style point of view: I’d actually put the whole check-and-exclude-caveats code in a function and have it run on both `devDependencies` and `dependencies` alike to keep it DRY.

File: test/dependency-checker.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+          return;
+        }
+
+        const caveatDetails = dependencyCaveat[dependencyName];
+        const dependencyDetails = deps[dependencyName];
+
+        if (
+          dependencyDetails.required === caveatDetails.overrideVersion ||
+          deps.gulp.stable === caveatDetails.currentVersion
```
Comment (surma) [inline test/dependency-checker.js]: This shouldn’t be `deps.gulp`, right?

**Pull request merged**